### ISSUE TITLE
Issue with GCC 5.5.0

### ISSUE THREAD

**Issue by navdeeprana** [ComputationalRadiationPhysics/picongpu#2851]
Installation:
`spack install picongpu backend=cuda ^cuda@9.0.176 ^openmpi@3.1.3 %gcc@5.5.0`

If you try to build LaserWakefield example, it fails with tons of error of the form
`error: argument of type "void *" is incompatible with parameter of type "xxxx *"`

There's a solution given at 
 [https://stackoverflow.com/questions/50815093/avx512-intrinsics-header-produces-many-errors-after-distro-upgrades-gcc-to-5-5-0](url)

And after trying the patch, atleast LaserWakefield example compiles and runs.
I am opening an issue, since it's docs mention to try installing with `GCC@5.5.0` with `cuda@9.0:1`, but this bug is not mentioned.
 
The gcc bug link : [https://gcc.gnu.org/bugzilla/show_bug.cgi?id=76731](url)
Forum post : [https://www.linuxquestions.org/questions/slackware-14/a-problem-about-gcc-5-5-0-after-upgrade-it-does-not-compile-tensorflow-4175623391/](url)

**Comment by ax3l** [comment]
Thank you for already looking up the issue, that is good to know!

Looks really like a GCC 5.5 issue on AVX512 (Skylake, Xeon Phi, ...). The best way to fix this is to really use a different version of GCC or not to build with AVX512 support when building with CUDA, where the heavy lifting is done anyway on the GPU and not the host-side.

Unfortunately, I don't think there is much we can do with the GCC issue. I am not even sure we can detect it easily to warn the user about/abort in configuration already.

For you to progress, I think the best thing to continue is with #2850 by either
- installing the latest CUDA 9.2 drivers on your host-system and go with a recent GCC or
- modifying the used boost version and stay with GCC 5.3.0

**Comment by navdeeprana** [comment]
> Unfortunately, I don't think there is much we can do with the GCC issue. I am not even sure we can detect it easily to warn the user about/abort in configuration already.
Maybe we can update the documentation, so atleast users are not caught unaware.
> Update: oh wait, of course if you manually fix the installed GCC 5.5.0 as you did with patches this works as well. Thank you for documenting!
Yes, this can be mentioned in the docs as well. I will be happy to do it.

**Comment by ax3l** [comment]
> Maybe we can update the documentation, so atleast users are not caught unaware.

Yes, I would love to but from the links it also looks like this actually was fixed from 5.4.0 to 5.5.0 and maybe this is just problem with the installed GCC. Did you install GCC 5.5.0 with Spack as well? Could it be possible something was mixed and the build dir was not purged before switching between the 5.3.0 and 5.5.0 versions you tried?

> Yes, this can be mentioned in the docs as well. I will be happy to do it.

We can certainly document the work-around you applied here, if this is still an issue with GCC 5.5.0. Please feel free to open a PR for the docs then, thanks!

**Comment by ax3l** [comment]
When switching from 5.3.0 to GCC 5.5.0, did you:
- open a new/fresh terminal to load the software again and
- remove the build directory?

In your input directory for an example:
```bash
rm -rf .build
```

It might be the headers were mixed up because of some leftover environment vars or some leftover CMake build files.

**Comment by navdeeprana** [comment]
> Did you install GCC 5.5.0 with Spack as well?

No the GCC 5.5.0 installation is from standard ubuntu repositories.

> We can certainly document the work-around you applied here, if this is still an issue with GCC 5.5.0. 
Please feel free to open a PR for the docs then, thanks!

It still is with the GCC 5.5.0 available on ubuntu repos.

> When switching from 5.3.0 to GCC 5.5.0, did you:
> 
>     * open a new/fresh terminal to load the dependencies again and
> 
>     * remove the build directory?

I recreated the entire myLWFA again. ;)

I read [https://www.linuxquestions.org/questions/slackware-14/a-problem-about-gcc-5-5-0-after-upgrade-it-does-not-compile-tensorflow-4175623391/](url) again and it mentions that the solution mentioned is not a very good idea, and probably it might be a cuda bug mentioned at [https://github.com/tensorflow/tensorflow/issues/10220#issuecomment-316887593](url)
I don't know about inner workings of compilers and cuda so I am not sure if this workaround should be mentioned at all.

**Comment by ax3l** [comment]
The CUDA issue is independent: CUDA 8.0 did not properly support GCC > 5.3: https://gist.github.com/ax3l/9489132 which is ill-documented by Nvidia but checked by us at configure time.

> I recreated the entire myLWFA again. ;)

Excellent, just needed to check because this is tricky sometimes with existing temporary build dirs.

Since this bug should be fixed in GCC 5.5 I do wonder if it's a Distro issue...

**Comment by BenjaminW3** [comment]
This should be the same issue we found in when testing alpaka: https://github.com/ComputationalRadiationPhysics/alpaka/issues/720
As far as I understood the bug descriptions, it was newly introduced into gcc 5.5 and was not available in any other version before or after.
So you either have to do a debug build or use a different CUDA or gcc version.
